### ISSUE TITLE
When asserting value returned from closure, getting null instead of the mocked value

### ISSUE THREAD

**Issue by werne2j** [ExpediaGroup/jenkins-spock#24]
I have a method that returns a closure that calls another method to get a string value and then return that back...
```groovy
def getOpenshiftNamespace() {
    return openshift.withCluster() {
        openshift.project() <--- this returns "Test Project"
    }
}
```
When writing a test for this, i am having trouble getting the value out of the closure when mocking the `openshift.project()` method. I have the value inside the closure but it does not get returned out and assigned to the variable. This is what i have as far as a test. Seems like it should be a straight forward test..
```groovy
  def "[getOpenshiftNamespace] will get openshift namespace" () {
    when:
      def namespace = utils.getOpenshiftNamespace()
    then:
      assert namespace == "Test Project" <--- namespace is null
      1 * getPipelineMock("openshift.withCluster")(_ as Closure)
      1 * getPipelineMock("openshift.project")() >> "Test Project"
  }
```
I can get the value if i return a string with `openshift.withCluster` such as 
```groovy
1 * getPipelineMock("openshift.withCluster")() >> {
    return "Test Project"
}
```
however i want to test the whole method and not just the outer layer.

**Comment by awittha** [comment]
What is `openshift`? Is it a global variable from a Jenkins plugin?

**Comment by werne2j** [comment]
Yes. I added it as a dependency so i am able to do `getPipelineMock("openshift")`

**Comment by awittha** [comment]
Thank you for your patience. Try this construction:

This Works
=====

```groovy
def "[getOpenshiftNamespace] will get openshift namespace" () {
		when:
			def namespace = Jenkinsfile.getOpenshiftNamespace()
		then:
			1 * getPipelineMock("openshift.project")()
			1 * getPipelineMock("openshift.withCluster")(_ as Closure) >> "Test Project"
			assert namespace == "Test Project"
	}
```

Note:

1. `openshift.project()` is not stubbed
2. `openshift.withCluster()` returns the "Test Project" string.

Why Does That Work?
=====

Why? Well, both `openshift.withCluster()` and `openshift.project` are mocks. In `jenkins-spock`, [mocks whose last argument is a closure will always execute that closure (because that's what most people probably want)](https://github.com/homeaway/jenkins-spock/blob/8c07bb245e758ca2dad2dcee574a2e1c38ce0c74/src/main/groovy/com/homeaway/devtools/jenkins/testing/JenkinsPipelineSpecification.groovy#L515-L520).... *but*, those mocks will always return their own, mocked self, _not_ the result of the closure.

So, when looking at code like this:

```groovy
def namespace = openshift.withCluster() {
    openshift.project()
}
```

`namespace` will be whatever  the`openshift.withCluster()` mock returns, which will be `null` unless you stub `openshift.withCluster()`.

The generic form for stubbing a mock to return the value of executing its "body" is

(one argument):
```groovy
getPipelineMock("node")(_ as Closure) >> { args -> args[0]() }
// e.g. node { someCode() }
```

(multiple arguments):
```groovy
getPipelineMock("stage")(_, _ as Closure) >> { args -> args[0].last()() }
// e.g. stage("name") { someCode() }
```

However, if you use that construction manually here, you'll actually run `openshift.project()` _twice_ - once when it's auto-executed, and once when it's executed as part of your stub...

Why Is It That Way?
=====

`jenkins-spock` decided to go with the automatic execution that doesn't return, for a couple reasons. The first two are succinct:

1.  So that users don't have to fill their unit tests with explicit "execute and return the body" lines for every single pipeline step with a body that their code might use somewhere.
2. If those mocks returned the result of the closure instead of their own mocked self, it would not be possible to stub return values for the mock interactions.

While the construction that works isn't the one you initially asked for, I would assert it's _cleaner_: Now, you assume that `openshift.withCluster()` works, and that `openshift.project()` should happen once, and these are *separate* assumptions/assertions.

I don't know exactly what `openshift.withCluster()` does, but you could imagine that not every pipeline step would always treat every body exactly the same. Some might conditionally execute it, some might pre- or post-process the body before returning it, and this would be encoded in the Java implementation isn't running (because the whole step is mocked) during the test-run. There might be some environment variables in scope inside the body block, added by the step.

In order to get a correct/expected value as a result of executing and returning the mocked step's body, you have to actually know and do all those things. Since the test execution is using a mock, you would have to manually stub out all that _behavior_. (this is the "third reason.")

Instead, you just stub the _expected result_, which is how mocking is supposed to work. It doesn't matter _what_ you write inside `openshift.withCluster()` so long as it returns the correct string "Test Project", so just mock that.

How Can I Test Both Mocks?
=====

If you want to test the inner workings, break them out into their own method and write separate tests for that, e.g.

`Utils.groovy`
```groovy
def getOpenshiftNamespace() {
    return openshift.withCluster() {
        return getOpenshiftProject()
    }
}

def getOpenshiftProject() {
	return openshift.project()
}
```

`UtilsSpec.groovy`
```groovy
def "[getOpenshiftNamespace] will get openshift namespace" () {
	when:
		def namespace = Util.getOpenshiftNamespace()
	then:
		1 * getPipelineMock("openshift.withCluster")(_ as Closure) >> "Test Project"
		assert namespace == "Test Project"
}

def "[getOpenshiftProject] will get openshift project" () {
	when:
		def project = Util.getOpenshiftProject()
	then:
		1 * getPipelineMock("openshift.project")() >> "Test Project"
		assert project == "Test Project"
		// 1 * someOtherImportantSideEffect()
		// 0 * sideEffectThatShouldntHappen()
}
```

In Short
=====

The original unit test was trying to test the implementation of `openshift.withCluster()`, but that is a mock and has no implementation. Stub the expected result of the whole computation, instead of stubbing each of the inner pieces - because a _mock_ execution won't necessarily do the same things the real step does, anyway.

Does this help?

**Comment by werne2j** [comment]
Thank you so much for a great explanation!

This makes perfect sense and was a great help for the problem i was having. 
I used a combination of separating out logic for certain methods (because i thought the logic 
needed to be tested more thoroughly) and mocking the outer closure to return a value such
as the first example you wrote. 

I believe this issue can be closed and hopefully is a help for anyone else who may come across this issue.

**Comment by deblaci** [comment]
Hi,

I also run into in this issue but, I don't want to stub the closure content, but I need the return value. There are some cases when not acceptabel to run twice. For WA I can move the "return" out of the closure but I don't like it :)

Could you please suggest other solution?

Eg.:
```
dir("test dir") {
  //do some stuff
  return "value"
} 
```

```
String incrementVersion(String v) {
  stages.stage("Increment version") {
    //do some stuff
    return v
  }
}
```
